command/remote_config_test.go

```diff
@@ -406,6 +406,128 @@ func TestRemoteConfig_enableRemote(t *testing.T) {
 	testRemoteLocalBackup(t, true)
 }
 
+func TestRemoteConfig_conflictRemote(t *testing.T) {
+	tmp, cwd := testCwd(t)
+	defer testFixCwd(t, tmp, cwd)
+
+	s := terraform.NewState()
+	s.Serial = 5
+	// Conflict is detected only when state has resources
+	s.Modules = []*terraform.ModuleState{
+		{
+			Path: []string{"root"},
+			Resources: map[string]*terraform.ResourceState{
+				"test_instance.foo": {},
+			},
+		},
+	}
+	conf, srv := testRemoteState(t, s, 200)
+	defer srv.Close()
+
+	s = s.DeepCopy()
+	// Different lineage locally means this is a conflict, because remote
+	// already exists but is describing an unrelated state lineage.
+	s.Lineage = "different-lineage"
+	s.Remote = conf
+
+	// Store the local state
+	statePath := filepath.Join(tmp, DefaultDataDir, DefaultStateFilename)
+	if err := os.MkdirAll(filepath.Dir(statePath), 0755); err != nil {
+		t.Fatalf("err: %s", err)
+	}
+	f, err := os.Create(statePath)
+	if err != nil {
+		t.Fatalf("err: %s", err)
+	}
+	err = terraform.WriteState(s, f)
+	f.Close()
+	if err != nil {
+		t.Fatalf("err: %s", err)
+	}
```

Comment:
The helper `testStateFile` should be used here to save you a lot of lines and cleanup. :)